hail/python/hailtop/batch/docs/cookbook/clumping.rst

```diff
@@ -0,0 +1,293 @@
+    def run_gwas(batch, vcf, phenotypes):
+        """
+        QC data and get association test statistics
+        """
+        cores = 2
+        gwas = batch.new_job(name='run-gwas')
+        gwas.image('gcr.io/<MY_PROJECT>/1kg-gwas:latest')
+        gwas.cpu(cores)
+        gwas.declare_resource_group(ofile={
+            'bed': '{root}.bed',
+            'bim': '{root}.bim',
+            'fam': '{root}.fam',
+            'assoc': '{root}.assoc'
+        })
+        gwas.command(f'''
+    python3 /run_gwas.py \
+        --vcf {vcf} \
+        --phenotypes {phenotypes} \
+        --output-file {gwas.ofile} \
+        --cores {cores}
+    ''')
+        return gwas
+
+A couple of things to note about this function:
+
+ - We specify the image to use is the image we created above. We copied the `run_gwas.py`
+   script into the root directory `/`. Therefore, to execute the `run_gwas.py` script, we
+   call `/run_gwas.py`.
+
+ - The `run_gwas.py` script takes an output-file parameter and then creates files ending with
+   the extensions `.bed`, `.bim`, `.fam`, and `.assoc`. In order for Batch to know the script is
+   creating files as a group with a common file root, we need to use the :meth:`.Job.declare_resource_group`
+   method. We then pass `gwas.ofile` as the output file root to `run_gwas.py` as that represents the temporary file
+   root given to all files in the resource group (`{root}`) when declaring the resource group).
+
+
+Clumping By Chromosome
+~~~~~~~~~~~~~~~~~~~~~~
+
+The second function performs clumping for a given chromosome. The input arguments are the :class:`.Batch`
+for which to create a new :class:`.Job`, the PLINK binary file **root**, the association results
+with at least two columns (SNP and P), and the chromosome for which to do the clumping for.
+The return value is the new :class:`.Job` created.
+
+.. code-block:: python
+
+    def clump(batch, bfile, assoc, chr):
+        """
+        Clump association results with PLINK
+        """
+        c = batch.new_job(name=f'clump-{chr}')
+        c.image('hailgenetics/genetics:0.2.37')
+        c.memory('1Gi')
+        c.command(f'''
+    plink --bfile {bfile} \
+        --clump {assoc} \
+        --chr {chr} \
+        --clump-p1 0.01 \
+        --clump-p2 0.01 \
+        --clump-r2 0.5 \
+        --clump-kb 1000 \
+        --memory 1024
+
+    mv plink.clumped {c.clumped}
+    ''')
+        return c
+
+A couple of things to note about this function:
+
+ - We use the image `'hailgenetics/genetics:0.2.37'` which is a publicly available Docker
+   image from Dockerhub maintained by the Hail team that contains many useful bioinformatics
+   tools including PLINK.
+
+ - We explicitly tell PLINK to only use 1Gi of memory because otherwise PLINK tries to
+   reserve half of the memory available on the machine which is not the same amount as what Batch
+   actually gives the job. Not specifying resource requirements correctly can cause performance
+   degradations with PLINK.
```

Comment:
I would phrase this differently:

> We explicitly tell PLINK to only use 1Gi of memory because PLINK defaults to using half of the machine's memory. PLINK's memory-available detection mechanism is unfortunately unaware of the memory limit imposed by Batch. ...